File: tests/algorithms/transpiler/FS/electronics/apparent_power.fs
// Generated 2025-08-14 09:59 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let PI: float = 3.141592653589793
let rec abs (x: float) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable x = x
    try
        __ret <- if x < 0.0 then (-x) else x
        raise Return
        __ret
    with
        | Return -> __ret
and to_radians (deg: float) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable deg = deg
    try
        __ret <- (deg * PI) / 180.0
        raise Return
        __ret
    with
        | Return -> __ret
and sin_taylor (x: float) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable x = x
    try
        let mutable term: float = x
        let mutable sum: float = x
        let mutable i: int = 1
        while i < 10 do
            let k1: float = 2.0 * (float i)
            let k2: float = k1 + 1.0
            term <- (((-term) * x) * x) / (k1 * k2)
            sum <- sum + term
            i <- i + 1
        __ret <- sum
        raise Return
        __ret
    with
        | Return -> __ret
and cos_taylor (x: float) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable x = x
    try
        let mutable term: float = 1.0
        let mutable sum: float = 1.0
        let mutable i: int = 1
        while i < 10 do
            let k1: float = (2.0 * (float i)) - 1.0
            let k2: float = 2.0 * (float i)
            term <- (((-term) * x) * x) / (k1 * k2)
            sum <- sum + term
            i <- i + 1
        __ret <- sum
        raise Return
        __ret
    with
        | Return -> __ret
and rect (mag: float) (angle: float) =
    let mutable __ret : float array = Unchecked.defaultof<float array>
    let mutable mag = mag
    let mutable angle = angle
    try
        let c: float = cos_taylor (angle)
        let s: float = sin_taylor (angle)
        __ret <- unbox<float array> [|mag * c; mag * s|]
        raise Return
        __ret
    with
        | Return -> __ret
and multiply (a: float array) (b: float array) =
    let mutable __ret : float array = Unchecked.defaultof<float array>
    let mutable a = a
    let mutable b = b
    try
        __ret <- unbox<float array> [|((_idx a (int 0)) * (_idx b (int 0))) - ((_idx a (int 1)) * (_idx b (int 1))); ((_idx a (int 0)) * (_idx b (int 1))) + ((_idx a (int 1)) * (_idx b (int 0)))|]
        raise Return
        __ret
    with
        | Return -> __ret
and apparent_power (voltage: float) (current: float) (voltage_angle: float) (current_angle: float) =
    let mutable __ret : float array = Unchecked.defaultof<float array>
    let mutable voltage = voltage
    let mutable current = current
    let mutable voltage_angle = voltage_angle
    let mutable current_angle = current_angle
    try
        let vrad: float = to_radians (voltage_angle)
        let irad: float = to_radians (current_angle)
        let vrect: float array = rect (voltage) (vrad)
        let irect: float array = rect (current) (irad)
        let result: float array = multiply (vrect) (irect)
        __ret <- result
        raise Return
        __ret
    with
        | Return -> __ret
and approx_equal (a: float array) (b: float array) (eps: float) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable a = a
    let mutable b = b
    let mutable eps = eps
    try
        __ret <- ((abs ((_idx a (int 0)) - (_idx b (int 0)))) < eps) && ((abs ((_idx a (int 1)) - (_idx b (int 1)))) < eps)
        raise Return
        __ret
    with
        | Return -> __ret
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

File: tests/algorithms/x/Java/electronics/apparent_power.java
public class Main {
    static double PI = (double)(3.141592653589793);

    static double abs(double x) {
        if ((double)(x) < (double)(0.0)) {
            return (double)(-x);
        }
        return (double)(x);
    }

    static double to_radians(double deg) {
        return (double)((double)((double)(deg) * (double)(PI)) / (double)(180.0));
    }

    static double sin_taylor(double x) {
        double term = (double)(x);
        double sum_1 = (double)(x);
        java.math.BigInteger i_1 = java.math.BigInteger.valueOf(1);
        while (i_1.compareTo(java.math.BigInteger.valueOf(10)) < 0) {
            double k1_1 = (double)((double)(2.0) * (double)((((Number)(i_1)).doubleValue())));
            double k2_1 = (double)((double)(k1_1) + (double)(1.0));
            term = (double)((double)((double)((double)(-term) * (double)(x)) * (double)(x)) / (double)(((double)(k1_1) * (double)(k2_1))));
            sum_1 = (double)((double)(sum_1) + (double)(term));
            i_1 = new java.math.BigInteger(String.valueOf(i_1.add(java.math.BigInteger.valueOf(1))));
        }
        return (double)(sum_1);
    }

    static double cos_taylor(double x) {
        double term_1 = (double)(1.0);
        double sum_3 = (double)(1.0);
        java.math.BigInteger i_3 = java.math.BigInteger.valueOf(1);
        while (i_3.compareTo(java.math.BigInteger.valueOf(10)) < 0) {
            double k1_3 = (double)((double)((double)(2.0) * (double)((((Number)(i_3)).doubleValue()))) - (double)(1.0));
            double k2_3 = (double)((double)(2.0) * (double)((((Number)(i_3)).doubleValue())));
            term_1 = (double)((double)((double)((double)(-term_1) * (double)(x)) * (double)(x)) / (double)(((double)(k1_3) * (double)(k2_3))));
            sum_3 = (double)((double)(sum_3) + (double)(term_1));
            i_3 = new java.math.BigInteger(String.valueOf(i_3.add(java.math.BigInteger.valueOf(1))));
        }
        return (double)(sum_3);
    }

    static double[] rect(double mag, double angle) {
        double c = (double)(cos_taylor((double)(angle)));
        double s_1 = (double)(sin_taylor((double)(angle)));
        return ((double[])(new double[]{(double)((double)(mag) * (double)(c)), (double)((double)(mag) * (double)(s_1))}));
    }

    static double[] multiply(double[] a, double[] b) {
        return ((double[])(new double[]{(double)((double)((double)(a[_idx((a).length, 0L)]) * (double)(b[_idx((b).length, 0L)])) - (double)((double)(a[_idx((a).length, 1L)]) * (double)(b[_idx((b).length, 1L)]))), (double)((double)((double)(a[_idx((a).length, 0L)]) * (double)(b[_idx((b).length, 1L)])) + (double)((double)(a[_idx((a).length, 1L)]) * (double)(b[_idx((b).length, 0L)])))}));
    }

    static double[] apparent_power(double voltage, double current, double voltage_angle, double current_angle) {
        double vrad = (double)(to_radians((double)(voltage_angle)));
        double irad_1 = (double)(to_radians((double)(current_angle)));
        double[] vrect_1 = ((double[])(rect((double)(voltage), (double)(vrad))));
        double[] irect_1 = ((double[])(rect((double)(current), (double)(irad_1))));
        double[] result_1 = ((double[])(multiply(((double[])(vrect_1)), ((double[])(irect_1)))));
        return ((double[])(result_1));
    }

    static boolean approx_equal(double[] a, double[] b, double eps) {
        return Math.abs((double)(a[_idx((a).length, 0L)]) - (double)(b[_idx((b).length, 0L)])) < (double)(eps) && Math.abs((double)(a[_idx((a).length, 1L)]) - (double)(b[_idx((b).length, 1L)])) < (double)(eps);
    }
    public static void main(String[] args) {
    }

    static int _idx(int len, long i) {
        return (int)(i < 0 ? len + i : i);
    }
}
